content/client-lib-development-guide/features.textile

```diff
@@ -1534,40 +1478,59 @@ enum StatsIntervalGranularity:
   DAY
   MONTH
 
+class DeviceDetails:
+  + getDeviceInfo() => io DeviceDetails
+  id: String
+  platform: DeviceType
+  type: DeviceType
+  clientId: String?
+  metadata: Dict<String, String>
+  updateTokenId: String
+  resetId(String)
+  resetUpdateToken() => io
```

Comment:
I'm not sure either, but given the discussion above, perhaps for this release:

* `activate` - activate device with or without callback responsible for registering the device with Ably.  If no callback is provided, activate will a) request a push notification target from the OS, b) store that target info locally, c) register the device automatically with Ably via the REST API assuming it has permissions, and will store the `localDeviceToken` locally that is received from Ably when registered.  If a callback is provided, then c) is not needed and it is the responsibility of the dev to do this typically through a request to their servers which in turn communicate with Ably.  Note that if activate is called again, and the device has a push registration ID, we need to decide what action is taken i.e. is a new token obtained or not?  Perhaps we'd need `reactivate` instead so that it's explicit.
* `reset` which generates a new `deviceId` and removes all locally stored info relating to push.  I am not sure why we would want to allow a developer to set the `deviceId` as it should be random, and if the `deviceId` has changed, then the token and push target effectively becomes invalid anyway, so to keep things simply why not just discard the lot.
* `deactivate` is the same as `reset` except the `deviceId` does not change and it either contacts Ably and unregisters itself, or optionally takes a callback (like `activate`).
* `updateLocalToken(token)` which allows a a local token string to be set for the client

So the changes as far as I can tell are:

* The client library does not automatically handle anything to do with resetting the local token, it simply takes a token string as an argument
* We need to make sure we have a REST API that can allow devices to be registered and returns a local device token that is stored with the device in Cassandra, can allow a device to be deleted (deactivated) and deletes all associated records, can allow a `localDeviceToken` to be reset by `POST`ing to an endpoint that returns a new `localDeviceToken` ID. The latter API needs `push-admin` rights or to be authenticated with the existing `localDeviceToken`